scipy/stats/_survival.py

```diff
@@ -1,38 +1,213 @@
+    ----------
+    points : ndarray
+        The point estimate of the cumulative distribution function (CDF) or its
+        complement, the survival function (SF), at unique values of the sample.
+    """
+    points: np.ndarray
+    # Exclude these from __str__
+    _x: np.ndarray = field(repr=False)  # points at which function is estimated
+    _n: np.ndarray = field(repr=False)  # number "at risk"
+    _d: np.ndarray = field(repr=False)  # number of "deaths"
+    _sf: np.ndarray = field(repr=False)  # survival function for var estimate
+    _kind: str = field(repr=False)  # type of function: "cdf" or "sf"
+
+    def __init__(self, x, points, n, d, kind):
+        self.points = points
+        self._x = x
+        self._n = n
+        self._d = d
+        self._sf = points if kind == 'sf' else 1 - points
+        self._kind = kind
+        f0 = 1 if kind == 'sf' else 0  # leftmost function value
+        f1 = 1 - f0
+        # fill_value can't handle edge cases at infinity
+        x = np.insert(x, [0, len(x)], [-np.inf, np.inf])
+        y = np.insert(points, [0, len(points)], [f0, f1])
+        # `or` conditions handle the case of empty x, points
+        self._f = interpolate.interp1d(x, y, kind='previous',
+                                       assume_sorted=True)
+
+    def confidence_interval(self, confidence_level=0.95, method='Greenwood'):
+        """Compute a confidence interval around the CDF/SF point estimate
+
+        Parameters
+        ----------
+        confidence_level : float, default: 0.95
+            Confidence level for the computed confidence interval
+
+        method : str, optional
+            Method used to compute the confidence interval. Options are
+            "Greenwood" and "Exponential Greenwood".
+
+        Returns
+        -------
+        ci : ``ConfidenceInterval``
+            An object with attributes ``low`` and ``high``: arrays of the
+            lower and upper bounds of the confidence interval at unique values
+            of the sample.
+
+        Notes
+        -----
+        Confidence intervals are computed according to the Greenwood formula or
+        the more recent "Exponential Greenwood" formula as described in [1]_.
+
+        References
+        ----------
+        .. [1] Sawyer, Stanley. "The Greenwood and Exponential Greenwood
+               Confidence Intervals in Survival Analysis."
+               https://www.math.wustl.edu/~sawyer/handouts/greenwood.pdf
+
+        """
+        methods = {'greenwood': self._greenwood_ci,
+                   'exponential greenwood': self._exp_greenwood_ci}
+
+        message = f"`method` must be one of {set(methods)}."
+        if method.lower() not in methods:
+            raise ValueError(message)
+
+        message = "`confidence_level` must be a scalar between 0 and 1."
+        confidence_level = np.asarray(confidence_level)[()]
+        if confidence_level.shape or not (0 <= confidence_level <= 1):
+            raise ValueError(message)
```

Comment:
I think we should think about a general way to check that. We are duplicating a lot of checks for that. (Not for this PR)